### PR TITLE
Direct users to WP Admin theme install if their site is on Business o…

### LINKED ISSUE #52710: Unable to update themes by uploading theme zip via Calypso
Since [WP v5.5](https://wordpress.org/support/wordpress-version/version-5-5/) it is possible to update a theme or plugin by re-uploading the zip file.  Before, this was not possible and a `Install problem: Theme already installed on site.` error message would appear.

As it turns out, this is still the behaviour when we try to upload a file to update an existing theme or plugin via Calypso to a Jetpack or Atomic site:

Screen Shot on 2021-05-10 at 11:57:13

One path would be to try and incorporate this into Calypo's theme upload flow. This would ultimately benefit not just WordPress.com sites, but all Jetpack sites.

If this cannot be fixed, this would be another point in favor of using wp-admin/themes.php over wordpress.com/themes for AT sites. 

Context: 3977270-zen

### PR DESCRIPTION
**Pull request by tjcafferkey**
#### Changes proposed in this Pull Request

* If a user is on a Business/Ecom plan then direct them to the WP Admin "Theme Install" screen instead of Calypso because users cannot update their Theme manually in Calypso but they are able to via WP Admin since WP 5.5

#### Testing instructions

* Open Calypso.live generated URL
* Go to `/themes/your-site.wordpress.com` on a site with a plan _below Business_ and click "Install theme" button to the right to manually upload a theme
* You should get sent to a Calypso screen which presents an upgrade nudge
* Go to `/themes/your-site.wordpress.com` on a site with a plan _Business or above_ and click "Install theme" button to the right to manually upload a theme
* You should get taken to `/wp-admin/theme-install.php`

Fixes https://github.com/Automattic/wp-calypso/issues/52710

### CODE REVIEW

File: client/state/sites/selectors/has-business-or-ecommerce-plan.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * External dependencies
+ */
+import { get } from 'lodash';
```
Comment (cpapazoglou) [inline client/state/sites/selectors/has-business-or-ecommerce-plan.js]: There is an ongoing effort in deprecating `lodash` p4TIVU-9Bf-p2. 
Can we use https://github.com/Automattic/wp-calypso/blob/82112916d5658967414013aaa1a2be8a89d3f10f/client/state/sites/selectors/get-site-plan-slug.js#L11 which ironically also uses `lodash` but at least will make the deprecation easier?

Comment (cpapazoglou) [reply client/state/sites/selectors/has-business-or-ecommerce-plan.js]: Basically, I am not sure whether creating a `siteHasBusinessOrEcommercePlan` selector is the best way forward. Since themes & plugins can only be installed on Atomic, isn't `isAtomic` selector https://github.com/Automattic/wp-calypso/blob/2de5e432fc90d3793d8bac4f8aa3e7a7dfb5432c/client/state/selectors/is-site-automated-transfer.js#L13 a better option?

Comment (tjcafferkey) [reply client/state/sites/selectors/has-business-or-ecommerce-plan.js]: >Since themes & plugins can only be installed on Atomic

You can also have a Free Atomic site, and in this case you should see the Calypso upgrade nudge which would not happen if I used the `isAtomic` selector?

File: client/my-sites/themes/theme-showcase.jsx [hunk 2]
```diff
@@ -47,6 +48,18 @@ import RecommendedThemes from './recommended-themes';
...
+function getInstallThemeSlug( siteSlug, hasBusinessOrEcommercePlan ) {
+	if ( siteSlug && hasBusinessOrEcommercePlan ) {
+		return `https://${ siteSlug }/wp-admin/theme-install.php`;
+	}
+
+	if ( siteSlug ) {
+		return `/themes/upload/${ siteSlug }`;
+	}
+
+	return '/themes/upload';
```
Comment (cpapazoglou) [inline client/my-sites/themes/theme-showcase.jsx]: Can we return early for `! siteSlug` case here?
```suggestion
	if ( siteSlug && hasBusinessOrEcommercePlan ) {
		return `https://${ siteSlug }/wp-admin/theme-install.php`;
	}

	if ( siteSlug ) {
		return `/themes/upload/${ siteSlug }`;
	}

	return '/themes/upload';
```
```suggestion
     if ( !siteSlug ) {
		'/themes/upload';
	}
	
	if ( hasBusinessOrEcommercePlan ) {
		return `https://${ siteSlug }/wp-admin/theme-install.php`;
	}

	return `/themes/upload/${ siteSlug }`;
```

**Review by cpapazoglou — CHANGES_REQUESTED** [review]
Thanks @tjcafferkey for picking this! It works as advertised 🎉 !

I have left some code specific comments which we can further discuss.

File: client/state/sites/selectors/has-business-or-ecommerce-plan.js (hunk 1 above)
Comment (cpapazoglou) [reply client/state/sites/selectors/has-business-or-ecommerce-plan.js]: Ahh, yes! That's correct! In that case I am still wondering whether we need a selector or we can bake it directly in the component. Have you figured out we may reuse this?

Comment (cpapazoglou) [reply client/state/sites/selectors/has-business-or-ecommerce-plan.js]: We also need the `isAtomic` on top, cause having a Business on WPCOM should still show the nudge (that will initiate the automatic transfer). Instead I got an error

Comment (tjcafferkey) [reply client/state/sites/selectors/has-business-or-ecommerce-plan.js]: I think the selector might be useful but we should rename it to something like `canUploadCustomThemesAndPlugins` which fundamentally checks the sites plan for this? This way we can use the selector for the same problem in the plugins section and if the ability for this feature expands beyond Business and Ecommerce plans we can just update the selector?

I am trying to shift the mindset away from Simple vs Atomic checks to plan checks.

Comment (tjcafferkey) [reply client/state/sites/selectors/has-business-or-ecommerce-plan.js]: >We also need the isAtomic on top

I think we should keep this check outside the selector. Let me make some amends.

**Comment by cpapazoglou** [comment]
Here is one more difficult problem:

- Select a Free site and go to `/themes/your-site.wordpress.com`
- You should be seeing the Calypso Theme install page which contains the nudge
- Click "Switch Site"
- Select an eCom site
- You are seeing the Calypso Theme install page, while you should be redirected to WP Admin

Maybe moving the logic to the Calypso Theme install page so that we redirect right before loading it will solve this problem.

SS 2021-05-12 at 16 26 05

**Comment by tjcafferkey** [comment]
>Maybe moving the logic to the Calypso Theme install page so that we redirect right before loading it will solve this problem.

Good spot. I'll do this also.

**Comment by tjcafferkey** [comment]
@cpapazoglou fixed those bugs, it was a little tricky as we again came across the assumption that all Atomic sites are on the Business plan. I've made some fixes that avoid a tonne of refactoring.

File: client/my-sites/themes/theme-showcase.jsx [hunk 3]
```diff
@@ -295,7 +309,7 @@ class ThemeShowcase extends React.Component {
 							className="themes__upload-button"
 							compact
 							onClick={ this.onUploadClick }
-							href={ siteSlug ? `/themes/upload/${ siteSlug }` : '/themes/upload' }
+							href={ getInstallThemeSlug( siteSlug, canUploadThemesOrPlugins ) }
```
Comment (cpapazoglou) [inline client/my-sites/themes/theme-showcase.jsx:312]: Isn't the redirection we do on the `Upload` component enough?
https://github.com/Automattic/wp-calypso/pull/52776/files#diff-579ae2c0c2c5ce152deb47027c130c06f93fabdf53b59dac04d8b15a69d60745R96-R98

**Comment by BogdanUngureanu** [comment]
Great job! I've tested it for lower plans and it works great (the user is redirected to the upgrade nudge), but it seems that for non-Atomic Business sites the https://site-name.wordpress.com/wp-admin/theme-install.php URL is broken. :( 

Screenshot 2021-05-13 at 18 34 55

**Comment by cpapazoglou** [comment]
> Great job! I've tested it for lower plans and it works great (the user is redirected to the upgrade nudge), but it seems that for non-Atomic Business sites the https://site-name.wordpress.com/wp-admin/theme-install.php URL is broken. :(
> 
> Screenshot 2021-05-13 at 18 34 55

On non-Atomic Business sites that link won't work until the site is transferred. How did you land in that page? In my tests I get the Calypso Upload theme screen which works as expected.

File: client/blocks/eligibility-warnings/utils.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export function isAtomicSiteWithoutBusinessPlan( holds: string[] ): boolean {
+	return holds.includes( 'TRANSFER_ALREADY_EXISTS' ) && holds.includes( 'NO_BUSINESS_PLAN' );
```
Comment (cpapazoglou) [inline client/blocks/eligibility-warnings/utils.ts:2]: Shouldn't we also check for `NO_ECOM_PLAN`?

File: client/blocks/eligibility-warnings/hold-list.tsx [hunk 5]
```diff
@@ -155,7 +156,12 @@ export const HardBlockingNotice = ( {
 	const blockingHold = holds.find( ( h ): h is keyof ReturnType< typeof getBlockingMessages > =>
 		isHardBlockingHoldType( h, blockingMessages )
 	);
-	if ( ! blockingHold ) {
+
+	/*
+		This accounts for Atomic sites that are on a plan below Business so we can prompt them to upgrade
+		instead of showing them an "Upload in progress" notice.
+	*/
+	if ( ! blockingHold || isAtomicSiteWithoutBusinessPlan( holds ) ) {
```
Comment (cpapazoglou) [inline client/blocks/eligibility-warnings/hold-list.tsx]: Can you please expand on why / where is that needed? Shouldn't we also check for eCommerce Plans and could we reuse `siteCanUploadThemesOrPlugins` selector?  

Also, this block renders in several places. Should it affect them?

**Comment by BogdanUngureanu** [comment]
> On non-Atomic Business sites that link won't work until the site is transferred. How did you land in that page? In my tests I get the Calypso Upload theme screen which works as expected.

I've followed the testing instructions on an non-Atomic site:
- My wordpress.com/me/account _Show wp-admin pages if available_ toggle is on
- Selected a non-Atomic site and clicked on _Appearance > Themes_ menu
- Was redirected to calypso-site-for-the-branch.com/themes/site-name.wordpress.com
- Clicked on the _Install theme_ button on the right of the page
- Was redirected to site-name.wordpress.com/wp-admin/theme-install.php which displays that 404 message.

**Comment by cpapazoglou** [comment]
> > On non-Atomic Business sites that link won't work until the site is transferred. How did you land in that page? In my tests I get the Calypso Upload theme screen which works as expected.
> 
> I've followed the testing instructions on an non-Atomic site:
> 
> * My wordpress.com/me/account _Show wp-admin pages if available_ toggle is on
> * Selected a non-Atomic site and clicked on _Appearance > Themes_ menu
> * Was redirected to calypso-site-for-the-branch.com/themes/site-name.wordpress.com
> * Clicked on the _Install theme_ button on the right of the page
> * Was redirected to site-name.wordpress.com/wp-admin/theme-install.php which displays that 404 message.

I am not sure why, but I can reliably reproduce it now. For me it's independent of `toggle` status. I always get redirected to WP Admin for non-Atomic business site.

**Comment by BogdanUngureanu** [comment]
> > > On non-Atomic Business sites that link won't work until the site is transferred. How did you land in that page? In my tests I get the Calypso Upload theme screen which works as expected.
> > 
> > 
> > I've followed the testing instructions on an non-Atomic site:
> > 
> > * My wordpress.com/me/account _Show wp-admin pages if available_ toggle is on
> > * Selected a non-Atomic site and clicked on _Appearance > Themes_ menu
> > * Was redirected to calypso-site-for-the-branch.com/themes/site-name.wordpress.com
> > * Clicked on the _Install theme_ button on the right of the page
> > * Was redirected to site-name.wordpress.com/wp-admin/theme-install.php which displays that 404 message.
> 
> I am not sure why, but I can reliably reproduce it now. For me it's independent of `toggle` status. I always get redirected to WP Admin for non-Atomic business site.

I've mentioned the toggle just to be sure that we can test in the same conditions (I haven't checked if it's reproduced based on the toggle).

File: client/blocks/eligibility-warnings/utils.ts (hunk 4 above)
Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/utils.ts:2]: There is no `NO_ECOM_PLAN`. If a site is on Atomic but is on a plan below Business it returns `NO_BUSINESS_PLAN` which means it needs to be on _at least_ a business plan.

File: client/blocks/eligibility-warnings/hold-list.tsx (hunk 5 above)
Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/hold-list.tsx]: >could we reuse `siteCanUploadThemesOrPlugins` selector

AFAIK no. The `siteCanUploadThemesOrPlugins` selector checks it has a business (or above) plan AND it's Atomic. We need to check that it's Atomic and a plan _below_ business.

Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/hold-list.tsx]: >Also, this block renders in several places. Should it affect them?

Based on the additional logic I've added to this, it's fine... I think.

**Comment by tjcafferkey** [comment]
>Was redirected to site-name.wordpress.com/wp-admin/theme-install.php which displays that 404 message.

Fixed in https://github.com/Automattic/wp-calypso/pull/52776/commits/f84316fa6382402796d63decf131bd3875de72ae

File: client/blocks/eligibility-warnings/hold-list.tsx (hunk 5 above)
Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/hold-list.tsx]: > Can you please expand on why / where is that needed?

Expanded on comments to explain usage in https://github.com/Automattic/wp-calypso/pull/52776/commits/3c581ce89ae2175f6b1a5b9054de293c779b5eb2

File: client/my-sites/themes/theme-showcase.jsx (hunk 3 above)
Comment (tjcafferkey) [reply client/my-sites/themes/theme-showcase.jsx:312]: It would work also, but I prefer to leave this because:

* We can avoid the user loading the additional view, when we are just going to redirect them. If we know their destination ahead of time the better.
* The intent feels more deliberate here too so anyone reading it will know what to expect.

File: client/my-sites/themes/theme-upload/index.jsx [hunk 6]
```diff
@@ -307,9 +320,11 @@ const mapStateToProps = ( state ) => {
 		progressLoaded: getUploadProgressLoaded( state, siteId ),
 		installing: isInstallInProgress( state, siteId ),
 		backPath: getBackPath( state ),
-		showEligibility: ! isJetpack && ( hasEligibilityMessages || ! isEligible ),
+		showEligibility:
+			( ! isJetpack && ( hasEligibilityMessages || ! isEligible ) ) || ! canUploadThemesOrPlugins,
```
Comment (mmtr) [inline client/my-sites/themes/theme-upload/index.jsx]: I don't think we want to show the eligibility warnings at all for Jetpack sites, since that's specific to the AT transfer
```suggestion
		showEligibility:
			( ! isJetpack && ( hasEligibilityMessages || ! isEligible || ! canUploadThemesOrPlugins ) ),
```

Anyway, I don't understand why is this needed. Why the ability to upload themes or plugins affects the AT transfer eligibility?

File: client/blocks/eligibility-warnings/index.tsx [hunk 7]
```diff
@@ -189,7 +190,11 @@ function getProceedButtonText( holds: string[], translate: LocalizeProps[ 'trans
 function isProceedButtonDisabled( isEligible: boolean, holds: string[] ) {
 	const resolvableHolds = [ 'NO_BUSINESS_PLAN', 'SITE_UNLAUNCHED', 'SITE_NOT_PUBLIC' ];
 	const canHandleHoldsAutomatically = holds.every( ( h ) => resolvableHolds.includes( h ) );
-	return ! canHandleHoldsAutomatically && ! isEligible;
+
+	// If it's not eligible for Atomic transfer lets also make sure it's not already Atomic with a plan below business.
+	return (
+		! canHandleHoldsAutomatically && ! isEligible && ! isAtomicSiteWithoutBusinessPlan( holds )
+	);
```
Comment (mmtr) [inline client/blocks/eligibility-warnings/index.tsx:197]: I think this is what's confusing me in my comment below.

Previously, the `EligibilityWarnings` was only handling the AT transfer (clicking on the "Proceed" button would always initiate the AT transfer) but now that doesn't seem to be case any longer? Are we using this component now as an upgrade nudge for sites already transferred to the AT infrastructure?

Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/index.tsx:197]: >Are we using this component now as an upgrade nudge for sites already transferred to the AT infrastructure?

Yes, AFAIK this component has always had the ability to show upgrade nudges. We have just adjusted the logic for our use case.

Comment (mmtr) [reply client/blocks/eligibility-warnings/index.tsx:197]: What if we keep the current behavior on the `EligibilityWarnings` component and do something like:

```js
if ( ! isAtomicSite ) {
  // render `EligibilityWarnings`
} else if ( isAtomicSiteWithoutBusinessPlan ) {
  // render `UpgradeNudge`
} 
```

Comment (mmtr) [reply client/blocks/eligibility-warnings/index.tsx:197]: > Yes, AFAIK this component has always had the ability to show upgrade nudges. We have just adjusted the logic for our use case.

Ok, that makes sense. What confuses me is that there is now a disconnection between the `isEligibleForAutomatedTransfer` selector and the `EligibilityWarnings` component, but maybe that's a not big deal. I'll have a deeper thought at this to see if there is a better alternative.

**Review by cpapazoglou — APPROVED** [review]
Thanks for addressing the feedback! It works as expected now in non-Atomic Business plans.

File: client/state/sites/selectors/can-upload-themes-or-plugins.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+ */
+import getSitePlanSlug from 'calypso/state/sites/selectors/get-site-plan-slug';
+import { isBusinessPlan, isEcommercePlan } from '@automattic/calypso-products';
+import isSiteWpcomAtomic from 'calypso/state/selectors/is-site-wpcom-atomic';
+
+// Only sites with plans Business and Ecommerce can install third-party themes and/or plugins.
+export default function siteCanUploadThemesOrPlugins( state, siteId ) {
+	const planSlug = getSitePlanSlug( state, siteId );
+	return (
+		( isBusinessPlan( planSlug ) || isEcommercePlan( planSlug ) ) &&
```
Comment (mmtr) [inline client/state/sites/selectors/can-upload-themes-or-plugins.js]: This could reuse the existing `isSiteOnAtomicPlan` selector: https://github.com/Automattic/wp-calypso/blob/trunk/client/state/selectors/is-site-on-atomic-plan.js

Comment (tjcafferkey) [reply client/state/sites/selectors/can-upload-themes-or-plugins.js]: Done

File: client/blocks/eligibility-warnings/index.tsx (hunk 7 above)
Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/index.tsx:197]: >What if we keep the current behavior on the EligibilityWarnings component and do something like:

Im not sure how you envisioned this but this button is de-coupled from the notices. Also both types of notices (Upgrades/Warnings/Blocks etc) are handled in the same component which is the [HoldList](https://github.com/Automattic/wp-calypso/blob/trunk/client/blocks/eligibility-warnings/hold-list.tsx)

File: client/my-sites/themes/theme-upload/index.jsx [hunk 9]
```diff
@@ -307,9 +322,12 @@ const mapStateToProps = ( state ) => {
 		progressLoaded: getUploadProgressLoaded( state, siteId ),
 		installing: isInstallInProgress( state, siteId ),
 		backPath: getBackPath( state ),
-		showEligibility: ! isJetpack && ( hasEligibilityMessages || ! isEligible ),
+		showEligibility:
+			( ( isJetpack && isAtomic ) || ( ! isJetpack && ! isAtomic ) ) &&
```
Comment (mmtr) [inline client/my-sites/themes/theme-upload/index.jsx]: Minor: all Atomic sites are Jetpack sites, so I think this can be simplified to:
```suggestion
			( isAtomic || ! isJetpack ) &&
```

File: client/my-sites/themes/theme-upload/index.jsx (hunk 6 above)
Comment (tjcafferkey) [reply client/my-sites/themes/theme-upload/index.jsx]: This is to show the eligibility, and this value also presents the user with the correct notice or nudge e.g. "Upgrade to business plan", "This will make your site Atomic" etc.

File: client/my-sites/themes/theme-upload/index.jsx (hunk 9 above)
Comment (tjcafferkey) [reply client/my-sites/themes/theme-upload/index.jsx]: Thank you. I was staring at this for too long to be able to simplify it. Done in https://github.com/Automattic/wp-calypso/pull/52776/commits/86feb063a09b0d642a52eefd2b5cc180250e30d5

Comment (tjcafferkey) [reply client/my-sites/themes/theme-upload/index.jsx]: I retested the logic and it works as expected still.

File: client/blocks/eligibility-warnings/hold-list.tsx [hunk 10]
```diff
@@ -155,7 +156,14 @@ export const HardBlockingNotice = ( {
...
 	);
-	if ( ! blockingHold ) {
+
+	/*
+		For Atomic sites on plans below Business it will return the holds TRANSFER_ALREADY_EXISTS and NO_BUSINESS_PLAN.
+		Because TRANSFER_ALREADY_EXISTS is present and 'blocking' it will show an "Upload in progress" notice even when there isn't one.
+		In this scenario we need to check if it's an Atomic ste (TRANSFER_ALREADY_EXISTS) on a plan below Business (NO_BUSINESS_PLAN)
+		so we can stop the render of "Upload in progress" and prompt them to upgrade instead.
+	*/
+	if ( ! blockingHold || isAtomicSiteWithoutBusinessPlan( holds ) ) {
```
Comment (mmtr) [inline client/blocks/eligibility-warnings/hold-list.tsx:166]: Minor and non-blocking: If this check is done within `isHardBlockingHoldType` you can avoid repeating the same check below. You'd need to change the signature of the `isHardBlockingHoldType` function so the list of holds is passed as an extra param, but I think this improves the readability.

Comment (tjcafferkey) [reply client/blocks/eligibility-warnings/hold-list.tsx:166]: The `isHardBlockingHoldType` function is also used in the `hasBlockingHold` function which is also imported elsewhere. I will see if I can understand this functions usages elsewhere with enough confidence to move this check but this was my original reason for keeping it separate.

File: client/my-sites/themes/theme-showcase.jsx [hunk 11]
```diff
@@ -47,6 +48,18 @@ import RecommendedThemes from './recommended-themes';
...
 import './theme-showcase.scss';
 
+function getInstallThemeSlug( siteSlug, canUploadThemesOrPlugins ) {
+	if ( ! siteSlug ) {
+		return '/themes/upload';
+	}
+
+	if ( canUploadThemesOrPlugins ) {
+		return `https://${ siteSlug }/wp-admin/theme-install.php`;
+	}
```
Comment (mmtr) [inline client/my-sites/themes/theme-showcase.jsx:58]: I found weird that after clicking on the "Install theme" button on a screen (Calypso) that is listing some themes, I'm taking away to a screen (WP Admin) that is again listing some themes where I need to click once more on the "Upload theme" button. 

Wouldn't that confuse users? Is there any way for automatically displaying the theme upload form when users land on `theme-install.php`?

Comment (mmtr) [reply client/my-sites/themes/theme-showcase.jsx:58]: Doesn't seem possible. The form is hidden by default, and it only shows up when clicking on the button: https://github.com/WordPress/WordPress/blob/51fe01e66de08e16e734b88af163ef80f4c4482e/wp-admin/js/theme.js#L1876-L1886

I guess we don't have any other alternative then.

**Review by mmtr — APPROVED** [review]
This tests well for me.
